serf/serf_test.go

```diff
@@ -836,6 +836,68 @@ func TestSerfRemoveFailedNode(t *testing.T) {
 	testMember(t, s3.Members(), s2Config.NodeName, StatusLeft)
 }
 
+func TestSerfRemoveFailedNode_prune(t *testing.T) {
+	s1Config := testConfig()
+	s2Config := testConfig()
+	s3Config := testConfig()
+
+	s1, err := Create(s1Config)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	s2, err := Create(s2Config)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	s3, err := Create(s3Config)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	defer s1.Shutdown()
+	defer s2.Shutdown()
+	defer s3.Shutdown()
+
+	_, err = s1.Join([]string{s2Config.MemberlistConfig.BindAddr}, false)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	_, err = s1.Join([]string{s3Config.MemberlistConfig.BindAddr}, false)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	testutil.Yield()
+
+	// Now force the shutdown of s2 so it appears to fail.
+	if err := s2.Shutdown(); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	time.Sleep(s2Config.MemberlistConfig.ProbeInterval * 5)
+
+	// Verify that s2 is "failed"
+	testMember(t, s1.Members(), s2Config.NodeName, StatusFailed)
+
+	// Now remove the failed node
+	if err := s1.RemoveFailedNode(s2Config.NodeName, true); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	// Check to make sure it's gone
+	if len(s1.Members()) != 2 {
```

Comment:
This seems like it would end up being super flaky given the timing involved. Is it worth using a retry helper around this and the assertion below to allow time for the force reap to propagate?

Reply:
I'm down for this as it may make it easier to do some test cleanup in the future. How would I go about it? import `"github.com/hashicorp/consul/sdk/testutil/retry"` or is there a way I put it in `"github.com/hashicorp/serf/testutil"`
🤔

Reply:
You can't have serf depend on consul because that would cause a not fun circular dependency. You could copy just the `retry` package as `github.com/hashicorp/serf/internal/retry` so that serf could use it, but nobody else could.